presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -446,8 +446,8 @@ public synchronized void finishInsertIntoExistingTable(
         setShared();
         SchemaTableName schemaTableName = new SchemaTableName(databaseName, tableName);
         Action<TableAndMore> oldTableAction = tableActions.get(schemaTableName);
-        if (oldTableAction == null) {
-            Table table = delegate.getTable(databaseName, tableName)
+        if (oldTableAction == null || oldTableAction.getData().getTable().isTemporary()) {
+            Table table = getTable(databaseName, tableName)
```

Comment:
No. By introducing this change we are allowing `INSERT_INTO` a newly created temporary table.

Reply:
@arhimondr I understand, I am saying why we remove `delegate` -- note previously when `oldTableAction == null`, we call `delegate.getTable` , but now we call `getTable` directly.

We can talk in person if it's more efficient :)

Reply:
Now i got what you are asking. 

Before the algorithm was like
```
if (oldTableAction == null) {
  delegate.getTable // get existing table from the metastore. As because of "oldTableAction == null" it is known that table is not buffered for the current transaction
  ...
  ...
  ... update existing table ...
  ... buffer updated table for the current transaction ...
}
```

Now, since a temporary table may be already buffered, we want to get it and update it's location. So now the algorithm is like

```
if (oldTableAction == nul || oldTableAction.isTemporary ) {
  getTable(databaseName, tableName) // get existing table either from the transaction buffer, or from the metastore. It can be in thransaction buffer only if it is temporary table
  ...
  ...
  ... update existing table ...
  ... buffer updated table for the current transaction ...
}
```

Reply:
So now we do the `tableActions` hash table lookup one more time , but doesn't change the behavior :) .